### ISSUE TITLE
[RFC] defmt-test: add unit test support

### ISSUE THREAD

**Issue by japaric** [knurling-rs/defmt#607]
## Summary

Extend `#[defmt_tests::tests]` to support "unit testing" (internal testing) an embedded no-std crate, i.e. with `cargo test --lib`

## Motivation

`defmt_test` currently supports "integration tests" (external testing) that live in `tests/*.rs`.
These tests (`#[test]` functions) can only test the public API of a (library) crate.

It would be advantageous to also support "internal testing" and reach feature parity with the standard `test` harness.
For example, this would let one add unit tests to a HAL and be able to organize those tests by module, where each module could map to a peripheral for example.

## Design

`#[defmt_test::tests]` can be used in any module and as many modules as one wants.

``` rust
// crate: my-hal

#![cfg_attr(test, no_main)] // <- REQUIRED
#![no_std]

mod serial {
    // ..

    #[defmt_test::tests]
    mod tests {
        #[test]
        fn it_works() {
            // ..
        }
        // ..
    }
}

mod i2c {
    // ..

    #[defmt_test::tests]
    mod tests {
        #[test]
        fn it_works() {
            // ..
        }
        // ..
    }
}

// top-level unit tests (these can overlap with `tests/*.rs` but are also supported)
#[defmt_test::tests]
mod tests {
    #[test]
    fn it_works() {
        // ..
    }
    // ..
}
```

To use the `defmt-test` test harness one needs to include the `#![cfg_attr(test, no_main)]` in the root of the crate and disable the standard test-harness in `Cargo.toml`:

``` toml
[lib]
harness = false # <- REQUIRED
```

With those changes in place, `cargo test --lib` will run all the unit tests in the crate:

> NOTE output format is not final

``` console
$ cargo t --lib
running (1/3) i2c::tests::it_works ...
running (2/3) tests::it_works ...
running (3/3) serial::tests::it_works ...
```

The order in which test *modules* are executed is left unspecified.

The tests within a single test module are executed in declaration order:

``` rust
#[defmt_test::tests]
mod tests {
    #[test]
    fn runs_first() {
        // ..
    }

    #[test]
    fn runs_second() {
        // ..
    }

    // .. later tests run afterwards
}
```

## Implementation details

This section details a possible implementation but the final implementation is free to diverge from what's described here.

TL;DR we use something like the `.init_array` "feature" used by ELF files for execution of functions before main (unsure if that's some sort of standard)

### `#[defmt_test::tests]`

We change the expansion of `#[defmt_test::tests]` to *not* include a `main` function.
Instead the `main` function is defined in `defmt_test` crate so that linking to `defmt_test` brings in the symbol.

Each `#[defmt_test::tests]` module will place a variable in the `.defmt_tests` linker section.
The static variable will contain a list of the `#[test]` functions defined in the module.

Given this input

``` rust
#[defmt_test::tests]
mod tests {
    #[init]
    fn init() -> State {
         // ..
    }

    #[test]
    fn foo(state: &mut State) {
        // .. user code ..
    }

    #[test]
    fn bar() {
        // .. user code ..
    }

    #[test]
    fn baz() -> Result<(), ()> {
        // .. user code ..
    }
}
```

The macro produces:

``` rust
#[cfg(test)]
mod tests {
    use defmt_test as _; // link in `defmt_test` to inject the `main` symbol

    #[linker_section = ".defmt_tests"]
    static TESTS: [extern "C" fn(); 4] = [
        init,
        foo,
        bar,
        baz,
    ];

    // trampoline that
    // - calls user `init`
    // - stores its return value in a static variable
    extern "C" fn init() {
        fn init() -> State {
            // .. user code ..
        }

        // ..
    }

    // trampoline that
    // - calls user `foo` with `State`
    // - reports progress using `defmt::println!`
    extern "C" fn foo() {
        fn foo(state: &mut State) {
             // .. user code ..
        }
        // ..
    }

    // similar trampolines
    extern "C" fn bar() { /* .. */ }
    extern "C" fn baz() { /* .. */ }
}
```

### linker script

in some linker script, maybe in `defmt.x`, we'll need something like this:

``` text
.defmt_tests : FLASH {
  _defmt_tests_start = ALIGN(4);
  KEEP(*(.defmt_tests));
  _defmt_tests_end = ALIGN(4);
}
```

### `main`

The `main` function in `defmt_tests` needs to do something like this:

``` rust
#[no_mangle]
extern "C" fn main() -> ! {
    extern "C" {
        static _defmt_tests_start: u32;
        static _defmt_tests_end: u32;
    }

    // execute all the trampolines in stored in `.defmt_tests`
    let mut cursor = &_defmt_tests_start as *const u32;
    let end = &_defmt_tests_end as *const u32;
    while cursor < end {
        unsafe {
            let trampoline: extern "C" fn() = mem::transmute(cursor);
            trampoline();
            cursor = cursor.offset(1);
        }
    }

    // make probe-run exit with exit-code = 0
    loop { asm::bkpt() }
}
```

## future extensions

these are outside the scope of this RFC and are only meant to explore some possibilities

- it could be useful to have some mechanism to reset all device peripherals at the start of each `#[defmt_test::tests]` module so that all the unit tests for a peripheral module start from a clean state.
  - it could even be necessary to software reset the device per test module to support existing HALs that use the "take once" singleton pattern as some peripherals, like the ones used in clock control, are needed by several high level peripheral APIs

**Comment by mhatzl** [comment]
This RFC looks like a feasible approach 👍 

Overall I really like the idea to extract `main` from the test-module.
Although it will definitely be a breaking change.

Some concerns and proposed changes from my side:

## Concerns with the RFC as it is proposed

- Handling multiple `init` functions

  At the moment, there can only be one `init` function per test-binary,
  because `defmt-test` only allows one in the test module.
  
  With the proposed change, each test-module could have its own `init` function.
  Essentially every test-module would be more like a group of tests, and a binary
  could now have multiple of such groups.
  This makes it difficult to run code that must only run once before executing all tests.
  The same goes for `teardown` functionality, because there is no guaranteed order in the execution of these groups.

- How to calculate the number of tests?

  Looking at the proc-macro code, the total number of tests is calculated by the number of tests found in the module.
  Not sure if the total number of tests added via `linker_section` can be tracked, because in the example above, the `init`, `before_each`, and `after_each` functions would also be part of the same section.

## Possible changes

Instead of putting `init` and `teardown` into the same section as the tests,
two additional linker sections should be used instead.
defmt-test could then fail if more than one function got linked.

`before_each` and `after_each` could be linked together with the tests.
This depends how the number of tests can be calculated.
Alternatively, wrapper functions could be created that internally call
`before_each`, `after_each`, and the actual test function.
I am just not sure if it is possible for defmt-test to get the correct modpath for a test function with this approach.

In case you additionally need a `init`/`teardown` function per module, I suggest renaming the module-scoped attributes to `before_all`/`after_all`.

These changes would require to create proc-macros for `init`, `teardown`, and `test`,
but would allow to directly annotate functions without the need to wrap them
in a defmt-test module.

Example:

```rust
// foo.rs
#[defmt_test::test]
fn test_foo() {}

// bar.rs
#[defmt_test::tests]
mod tests {
  #[before_all]
  fn before_all() {}

  #[before_each]
  fn before_each() {}

  #[after_each]
  fn after_each() {}

  #[after_all]
  fn after_all() {}

  #[test]
  fn test_bar1() {}

  #[test]
  fn test_bar2() {}
}

// lib.rs
use defmt_test as _;

#[defmt_test::init]
fn init() {}

#[defmt_test::teardown]
fn teardown() {}
```

This could allow the following flow:

1. `init()`
2. `test_foo()`
3. test group in bar.rs
   1. `before_all()`
   2. `before_each()`
   3. `test_bar1()`
   4. `after_each()`
   5. `before_each()`
   6. `test_bar2()`
   7. `after_each()`
   8. `after_all()`
4. `teardown()`

**Comment by Urhengulas** [comment]
- https://crates.io/crates/satchel
- https://crates.io/crates/linkme
